src/org/sosy_lab/java_smt/solvers/princess/PrincessStringFormulaManager.java

```diff
@@ -0,0 +1,312 @@
+        // The longest allowed literal is \\u{d5 d4 d3 d2 d1}
+        return pToken;
+      }
+      literal = pToken.substring(3, pToken.length() - 1);
+    } else {
+      if (pToken.length() != 6) {
+        // Abort if there are not exactly 4 digits
+        // The literal must have this form: \\u d3 d2 d1 d0
+        return pToken;
+      }
+      literal = pToken.substring(2);
+    }
+
+    // Try to parse the digits as a (hexadecimal) integer
+    // Abort if there is an error
+    int value;
+    try {
+      value = Integer.parseInt(literal, 16);
+    } catch (NumberFormatException e) {
+      return pToken;
+    }
+
+    // Return the Unicode letter if it fits into a single 16bit character
+    // Otherwise throw an exception as we don't support Unicode characters with more than 4 digits
+    char[] chars = Character.toChars(value);
+    if (chars.length != 1) {
+      throw new IllegalArgumentException();
+    } else {
+      return String.valueOf(chars[0]);
+    }
+  }
+
+  /** Replace escape sequences for Unicode letters with their UTF16 representation. */
+  static String unescapeString(String pInput) {
+    StringBuilder builder = new StringBuilder();
+    while (!pInput.isEmpty()) {
+      // Search for the next escape sequence
+      int start = pInput.indexOf("\\u");
+      if (start == -1) {
+        // Append the rest of the String to the output if there are no more escaped Unicode
+        // characters
+        builder.append(pInput);
+        pInput = "";
+      } else {
+        // Store the prefix up to the escape sequence
+        String prefix = pInput.substring(0, start);
+
+        // Skip ahead and get the escape sequence
+        pInput = pInput.substring(start);
+        String value;
+        if (pInput.charAt(2) == '{') {
+          // Sequence has the form \\u{d5 d4 d3 d2 d1 d0}
+          // Find the closing bracket for the literal:
+          int stop = pInput.indexOf('}');
+          if (stop == -1) {
+            // Use the index right after "\\u{" if there is no closing bracket
+            stop = 2;
+          }
+          value = pInput.substring(0, stop + 1);
+          pInput = pInput.substring(stop + 1);
+        } else {
+          // Sequence has the form \\u d3 d2 d1 d0
+          int stop = 2;
+          while (stop < pInput.length()) {
+            char c = pInput.charAt(stop);
+            if (Character.digit(c, 16) == -1) {
+              break;
+            }
+            stop++;
+          }
+          value = pInput.substring(0, stop);
+          pInput = pInput.substring(stop);
+        }
+
+        // Try to parse the escape sequence to replace it with its 16bit Unicode character
+        // If parsing fails just keep it in the String
+        String nextToken = literalOrSkip(value);
+
+        // Collect the prefix and the (possibly) translated escape sequence
+        builder.append(prefix).append(nextToken);
+      }
+    }
+    return builder.toString();
+  }
```

Comment:
Just a comment, not blocking the merge:
Does escaping and unescaping with Strings align with all operations like String comparison and String inclusions? This feature seems to be quite complex.

Reply:
Hmm, that's a good question..

I wrote this method since many/most of the tests in `StringFormulaManagerTest` use (SMTLIB) escaped Unicode, which Princess does not seem  to support. It should be fine for that purpose, but I'm not sure how well escaping/unescaping will work in general. This shouldn't be a problem for this pull request as we're currently only unescaping input Strings, but don't try to escape the output.

However, we might still want to have a closer look at this. To be honest I'm not even quite sure what the format of `value` in `StringFormulaManager.makeString(String value)` is supposed to be. Several of the tests use SMTLIB escape sequences here, but the argument is really just a regular Java String. Do we want allow Unicode characters in it or not? When I try `makeString("\\u{39E}")` different solver return different results: CVC5 and Princess recognize the escape sequence and treat it as `"Ξ"`, whereas CVC4 and Z3 do not and simply return `\u{39E}` again. On the other hand `makeString("Ξ")` will throw and exception for CVC4/5 (because of the invalid character), however it passes for Princess, and Z3 will read the String as `\u{ce}\u{9e}`, which is the UTF-8 encoding of `"Ξ"` and doesn't seem quite right either.

Maybe it would be best to figure this out as part of another issue and then come up with a common definition? In the mean time I may also ask the Ostrich developers to add support for escape Unicode characters as they are part of the SMTLIB standard.

Reply:
I opened a new issue for that: https://github.com/sosy-lab/java-smt/issues/412

Feel free to add more details there.